core/src/main/java/io/fabric8/maven/core/service/PortForwardService.java

```diff
@@ -0,0 +1,256 @@
+ * permissions and limitations under the License.
+ */
+package io.fabric8.maven.core.service;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import io.fabric8.kubernetes.api.KubernetesHelper;
+import io.fabric8.kubernetes.api.PodStatusType;
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodList;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.FilterWatchListDeletable;
+import io.fabric8.maven.core.util.KubernetesClientUtil;
+import io.fabric8.maven.core.util.KubernetesResourceUtil;
+import io.fabric8.maven.core.util.ProcessUtil;
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.utils.Strings;
+
+/**
+ * A service for forwarding connections to remote pods.
+ *
+ * @author nicola
+ * @since 28/03/2017
+ */
+public class PortForwardService {
+
+    private ClientToolsService clientToolsService;
+
+    private Logger log;
+
+    private KubernetesClient kubernetes;
+
+    public PortForwardService(ClientToolsService clientToolsService, Logger log, KubernetesClient kubernetes) {
+        this.clientToolsService = Objects.requireNonNull(clientToolsService, "clientToolsService");
+        this.log = Objects.requireNonNull(log, "log");
+        this.kubernetes = Objects.requireNonNull(kubernetes, "kubernetes");
+    }
+
+    /**
+     * Forwards a port to the newest pod matching the given selector.
+     * If another pod is created, it forwards connections to the new pod once it's ready.
+     */
+    public Closeable forwardPortAsync(final Logger externalProcessLogger, final LabelSelector podSelector, final int remotePort, final int localPort) throws Fabric8ServiceException {
+
+        final Lock monitor = new ReentrantLock(true);
+        final Condition podChanged = monitor.newCondition();
+        final Pod[] nextForwardedPod = new Pod[1];
+
+        final Thread forwarderThread = new Thread() {
+            @Override
+            public void run() {
+
+                Pod currentPod = null;
+                Closeable currentPortForward = null;
+
+                try {
+                    monitor.lock();
+
+                    while (true) {
+                        if (podEquals(currentPod, nextForwardedPod[0])) {
+                            podChanged.await();
+                        } else {
+                            Pod nextPod = nextForwardedPod[0]; // may be null
+                            monitor.unlock();
```

Comment:
It's just to exit the critical section since I'm doing possibly time-consuming tasks, to avoid blocking the watcher thread while trying to acquire the lock..